emacsPackages.cask: 0.9.0 -> 0.9.1

magicquark opened the pull request:
### Update and fix `emacsPackages.cask`

#### Update from 0.9.0 -> 0.9.1

Changelog:
https://github.com/cask/cask/compare/v0.9.0...v0.9.1

#### To Resolve https://github.com/NixOS/nixpkgs/issues/445465 (if backported to 25.05)

- Cask expects a local `package-build` directory to be present in the output, as it loads a custom version during bootstrap. See: https://github.com/cask/cask/blob/master/cask-bootstrap.el#L45

- The vendored version of `package-build` is then used to download from `melpa` and `elpa` the dependencies that `cask` requires if they do not exist already. This is contrary to the philosophy of `nixpkgs` whereby the dependencies should be sourced from `nixpkgs` itself.

- To avoid this, the bootstrap process has been patched to provide the required dependencies using `nixpkgs`.

- Additionally, `cask-cli` has a function `upgrade-cask` that uses the `cask-bootstrap-dir` that its own bootstrap process creates. See https://github.com/cask/cask/blob/fd16ea95664b8f13c55b8608dea63942d431ca00/cask-cli.el#L131. As the bootstrap process has been modified to use dependencies from `nixpkgs`, `cask` can no longer update itself. Therefore, this function has been patched to make it effectively a noop.

- These patches cannot be upstreamed as it is specific to the way that packages are integrated into `nixpkgs`.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -51,9 +53,9 @@ melpaBuild (finalAttrs: {
   postInstall = ''
-    install -D -t $out/bin bin/cask
+    installBin bin/cask
+    cp -R ${finalAttrs.src}/package-build $out/share/emacs/site-lisp/elpa/cask-${finalAttrs.melpaVersion}/
```
Your patch uses vendored `package-build` and installs all other dependencies of `cask` using `package.el`, which strictly follows upstream `cask` and should work (though I have not tested it).

However, as a distribution, we prefer not vendoring dependencies.  We should try to make `cask` use `package-build` and other dependencies from Nixpkgs when possible.  After skimming through `cask` source code, I think this can be done by replacing `cask-bootstrap.el` with something like this:

1. add all dependencies of `cask`, such as `f` and `package-build`, to `load-path`
2. `require` all dependencies

---

That being said, I am fine with merging your patch since it is an improvement of the current state.

jian-lin replied:
> add all dependencies of cask from Nixpkgs

To be clear, all dependencies refer to the dependency **closure**, i.e., not only the direct dependencies of cask.

magicquark replied:
Noted. I have revised the PR to provide the requirements from Nixpkgs, as suggested.

magicquark commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -51,9 +53,9 @@ melpaBuild (finalAttrs: {
   postInstall = ''
-    install -D -t $out/bin bin/cask
+    installBin bin/cask
+    cp -R ${finalAttrs.src}/package-build $out/share/emacs/site-lisp/elpa/cask-${finalAttrs.melpaVersion}/
```
When I tried to remove the `packageRequires` the package would not build. I have since revised the PR, so this point is now moot.

magicquark commented:
> Looks good. Is it ready for review?

Thank you for the initial review and comments.

I've updated the PR as suggested, now using Nixpkgs to manage `cask`'s dependencies. To ensure all dependencies are included in the closure, I implemented a recursive function.

I've tested the package locally, and its basic functionality appears to be working as expected.

It is ready for review.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -1,35 +1,56 @@
+  load-path-mod = builtins.concatStringsSep "\n" (
+    map (
+      x:
+      "(add-to-list 'load-path \\\"${x.outPath}/share/emacs/site-lisp/elpa/${x.pname}-${x.version}/\\\")"
```
I would try to keep `load-path-mod` simple, i.e., only generating data on the nixlang side and putting the logic of extending `load-path` in the patch.

magicquark replied:
I have revised this as you suggested. I think that it is better now.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -1,35 +1,56 @@
+      indirect = builtins.concatLists (map getAllDependenciesOfPkg direct);
+    in
+    lib.unique (direct ++ indirect);
+  dependencies = getAllDependenciesOfPkg cask;
```
`cask` should be replaced with `finalAttrs.finalPackage`

magicquark replied:
Done.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -47,13 +68,16 @@ melpaBuild (finalAttrs: {
 
   # use melpaVersion so that it works for unstable releases too
   postPatch = ''
-    lispdir=$out/share/emacs/site-lisp/elpa/cask-${finalAttrs.melpaVersion} \
-      substituteAllInPlace bin/cask
+    export lispdir=$out/share/emacs/site-lisp/elpa/cask-${finalAttrs.melpaVersion}
+    substituteAllInPlace bin/cask
```
I prefer not including unrelaed/unnecessary change

magicquark replied:
Noted, reverted.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -47,13 +68,16 @@ melpaBuild (finalAttrs: {
+    export lispdir=$out/share/emacs/site-lisp/elpa/cask-${finalAttrs.melpaVersion}
+    substituteAllInPlace bin/cask
+
+    substituteInPlace cask-bootstrap.el \
+      --replace "@deps-mod@" "${deps-mod}" \
+      --replace "@load-path-mod@" "${load-path-mod}"
```
It is better to use `replaceVars` to construct the patch first. See `lsp-bridge`'s nix expression for an example.

magicquark replied:
I see. Change made.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/0001-cask-bootstrap.diff:
We do AOT native compilation for elisp files.  Besides `load-path`, it would be good to also extend `native-comp-eln-load-path`.

magicquark replied:
I have added `native-comp-eln-load-path`, but I have been unsuccessful at adding the native dir of `cask` to it because of infinite recursion occurring.

magicquark replied:
I see. I think that I have been able to add `cask`'s native dir now. I have added the changes as another commit to make it obvious what changes have been made.

jian-lin commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -47,13 +68,16 @@ melpaBuild (finalAttrs: {
 
   # use melpaVersion so that it works for unstable releases too
   postPatch = ''
-    lispdir=$out/share/emacs/site-lisp/elpa/cask-${finalAttrs.melpaVersion} \
-      substituteAllInPlace bin/cask
+    export lispdir=$out/share/emacs/site-lisp/elpa/cask-${finalAttrs.melpaVersion}
+    substituteAllInPlace bin/cask
+
+    substituteInPlace cask-bootstrap.el \
+      --replace "@deps-mod@" "${deps-mod}" \
+      --replace "@load-path-mod@" "${load-path-mod}"
   '';
 
-  # TODO: use installBin as soon as installBin arrives Master branch
   postInstall = ''
-    install -D -t $out/bin bin/cask
+    installBin bin/cask
```
nit: could use an extra commit for this

magicquark committed:
emacsPackages.cask: replace install with installBin

magicquark committed:
emacsPackages.cask: 0.9.0 -> 0.9.1

Changelog:
https://github.com/cask/cask/compare/v0.9.0...v0.9.1

magicquark committed:
emacsPackages.cask: fix import error and use non-vendored dependencies

- Fixes issue 445465.

- Cask expects a local `package-build` directory to be present in the
  output, as it loads a custom version during bootstrap. See:
  https://github.com/cask/cask/blob/master/cask-bootstrap.el#L45

- The vendored version of `package-build` is then used to download from
  `melpa` and `elpa` the dependencies that `cask` requires. This is
  contrary to the philosophy of `nixpkgs` whereby the dependencies
  should be sourced from `nixpkgs` itself.

- To avoid this, the bootstrap process has been patched to provide the
  required dependencies using `nixpkgs`.

- Additionally, `cask-cli` has a function `upgrade-cask` that uses the
  `cask-bootstrap-dir` that its own bootstrap process creates. See
  https://github.com/cask/cask/blob/fd16ea95664b8f13c55b8608dea63942d431ca00/cask-cli.el#L131.
  As the bootstrap process has been modified to use dependencies from
  `nixpkgs`, `cask` can no longer update itself. Therefore, this
  function has been patched to make it effectively a noop.

- The patch cannot be upstreamed as it is specific to the way that
  packages are integrated into `nixpkgs`.

magicquark commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/0001-cask-bootstrap.diff:
```diff
@@ -0,0 +1,48 @@
+ 
+-(require 'package)
+-
+ (defvar cask-directory)
```
Removed.

magicquark commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/0001-cask-bootstrap.diff:
```diff
@@ -0,0 +1,48 @@
+-
+ (defvar cask-directory)
+ 
+ (defconst cask-bootstrap-dir
```
Removed, and I have patched the file "cask-cli.el" as it referenced the `cask-bootstrap-dir`.

magicquark commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/0001-cask-bootstrap.diff:
```diff
@@ -0,0 +1,48 @@
+-  (require 'package-build)
++;; Add nix store paths for dependencies to the load-path.
++@load-path-mod@
++(let ((deps '(@deps-mod@ cl-lib eieio)))
```
Ahh, I see. I have done as you proposed.

magicquark commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -1,35 +1,56 @@
+  load-path-mod = builtins.concatStringsSep "\n" (
+    map (
+      x:
+      "(add-to-list 'load-path \\\"${x.outPath}/share/emacs/site-lisp/elpa/${x.pname}-${x.version}/\\\")"
```
Updated.

jian-lin reviewed (changes requested):
Nice work!

magicquark commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/cask/package.nix:
```diff
@@ -4,68 +4,101 @@
+        inherit deps-mod load-path-mod native-load-path-mod;
+      })
+      # Disable function that uses cask-bootstrap-dir to update Cask in-place
+      ./0002-cask-cli-upgrade-cask-disable.diff
```
I have merged them.

magicquark commented:
> Nice work!

Thank you for the review and help.

jian-lin committed:
emacsPackages.cask: improve patch substitution

- use --replace-fail to fail fast
- use shell variable melpaVersion to avoid a string interperation
- use shell variable ename to be consistent with other code and to
  avoid hardcoding

jian-lin committed:
emacsPackages.cask: let cask use its own native compiled results

jian-lin reviewed (changes requested):
Looks good!  I think this PR is in good shape now and this will be my last comment before merging it!

jian-lin reviewed (approved):
LGTM, thanks!

I removed the prefix space of the commit message `emacsPackages.cask: fix import error and use non-vendored dependencies`.
